numpy/_core/src/umath/stringdtype_ufuncs.cpp

```diff
@@ -2142,6 +2144,171 @@ string_inputs_promoter(
+                     self->name);
+        return _NPY_ERROR_OCCURRED_IN_CAST;
+    }
+
+    for (int i = 0; i < 4; i++) {
+        Py_INCREF(given_descrs[i]);
+        loop_descrs[i] = given_descrs[i];
+    }
+
+    PyArray_StringDTypeObject *in_descr =
+            (PyArray_StringDTypeObject *)loop_descrs[0];
+    int out_coerce = in_descr->coerce;
+    PyObject *out_na_object = in_descr->na_object;
+    loop_descrs[4] = (PyArray_Descr *)new_stringdtype_instance(out_na_object,
+                                                               out_coerce);
+    if (loop_descrs[4] == NULL) {
+        return _NPY_ERROR_OCCURRED_IN_CAST;
+    }
+
+    return NPY_NO_CASTING;
+}
+
+static int
+slice_strided_loop(PyArrayMethod_Context *context, char *const data[],
+                   npy_intp const dimensions[], npy_intp const strides[],
+                   NpyAuxData *NPY_UNUSED(auxdata))
+{
+    char *iptr = data[0];
+    char *start_ptr = data[1];
+    char *stop_ptr = data[2];
+    char *step_ptr = data[3];
+    char *optr = data[4];
+
+    npy_intp N = dimensions[0];
+
+    npy_string_allocator *allocators[5] = {};
+    NpyString_acquire_allocators(5, context->descriptors, allocators);
+    npy_string_allocator *iallocator = allocators[0];
+    npy_string_allocator *oallocator = allocators[4];
+
+    // Build up an index mapping the codepoints to byte offsets in the encoded
+    // string.
+    std::vector<int> codepoint_offsets;
+
+    while (N--) {
+        // get the slice
+        npy_intp start = *(npy_intp *)start_ptr;
+        npy_intp stop = *(npy_intp *)stop_ptr;
+        npy_intp step = *(npy_intp *)step_ptr;
+
+        npy_static_string is = {0, NULL};
+        const npy_packed_static_string *ips = (npy_packed_static_string *)iptr;
+        npy_static_string os = {0, NULL};
+        npy_packed_static_string *ops = (npy_packed_static_string *)optr;
+        int is_isnull = NpyString_load(iallocator, ips, &is);
+        if (is_isnull == -1) {
+            npy_gil_error(PyExc_MemoryError, "Failed to load string in slice");
+            goto fail;
+        }
+        else if (is_isnull) {
+            npy_gil_error(PyExc_TypeError, "Cannot slice null string");
+            goto fail;
+        }
+
+        // determine number of codepoints in string
+        size_t num_codepoints = 0;
+        {
+            const char *inbuf_ptr = is.buf;
+            const char *inbuf_ptr_end = is.buf + is.size;
+
+            // ignore trailing nulls
+            while (inbuf_ptr < inbuf_ptr_end && *(inbuf_ptr_end - 1) == 0) {
+                inbuf_ptr_end--;
+            }
+
+            // count codepoints and build codepoint_offsets
+            codepoint_offsets.clear();
+            while (inbuf_ptr < inbuf_ptr_end) {
+                num_codepoints++;
+                int num_bytes = num_bytes_for_utf8_character(
+                        (const unsigned char *)inbuf_ptr);
+                codepoint_offsets.push_back(inbuf_ptr - is.buf);
```

Comment:
It occurs to me that you only need to store this if it isn't one byte. Since we assume ASCII/Latin-1 is the common case, we can probably get some performance back there by only storing offsets for non-Latin-1 bytes.

I think I can implement this today and just push to your branch.

I like using a `std::vector` here, and we can probably reduce the heap thrashing a lot for common scientific data just by implementing my above suggestion. I agree doing things with a thread-local here doesn't make sense.

Reply:
I think we can also compute `outsize` in this loop...

Reply:
I spent a little while on this and came up with the following design. Instead of using a vector of ints, you can use a vector of pairs of ints, where each pair is a codepoint index and offset into the array. Then you can compute outsize by first assuming that the size of the string is the number of codepoints in the slice (e.g. one byte per character) and then go through the offsets vec and for each multibyte character that is in the slice, add the extra bytes for that character to `outsize`. That avoids a pass over the whole string.

Reply:
On further reflection I don't think this is actually better than what you have here. Your approach is a lot simpler. I've refactored things a bit though and done some cleanups. I'm hoping to push something to your branch shortly.

Reply:
Thanks for looking into this! As you had alluded to in your first comment, do you want to take a stab at implementing that idea?

Regarding heap-thrashing: I was worried about this at first, but since the `std::vector` is reused across the loop over all array elements I think it is unlikely to be resized many times. For that you would need an array where every string is larger than all previous strings, I think.

Reply:
This ended up reducing down to a much smaller cleanup after I decided this approach was too complicated. I've marked this as resolved but please look over my last push.

Reply:
Looks good to me! I agree, your change simplifies the code.

I think it makes sense to keep this "simpler" implementation for now and worry about further optimizing performance only if really necessary. It's already massively faster for my use case:
```
>>> a = np.full(1_000_000, "😀😃😄😁😆😅🤣😂🙂🙃", dtype='T')
>>> start = time.time(); _ = np.array([s[3:7] for s in a], dtype=a.dtype); end = time.time()
>>> end - start
0.4303429126739502
>>> start = time.time(); _ = np.strings.slice(a, 3, 7); end = time.time()
>>> end - start
0.08485579490661621
```